src/ILVerify/src/ILImporter.Verify.cs

```diff
@@ -949,7 +957,34 @@ void ImportSwitchJump(int jmpBase, int[] jmpDelta, BasicBlock fallthrough)
 
         void ImportBranch(ILOpcode opcode, BasicBlock target, BasicBlock fallthrough)
         {
-            // TODO: Verification rules
+            switch (opcode)
+            {
+                case ILOpcode.br: break;
+                case ILOpcode.brfalse:
+                case ILOpcode.brtrue:
+                    {
+                        StackValue value = Pop();
+                        Check(value.Kind >= StackValueKind.Int32 && value.Kind <= StackValueKind.NativeInt || value.Kind == StackValueKind.ObjRef || value.Kind == StackValueKind.ByRef, VerifierError.StackUnexpected);
+                    }
+                    break;
+                case ILOpcode.beq:
+                case ILOpcode.bge:
+                case ILOpcode.bgt:
+                case ILOpcode.ble:
+                case ILOpcode.blt:
+                case ILOpcode.bne_un:
+                case ILOpcode.bge_un:
+                case ILOpcode.bgt_un:
+                case ILOpcode.ble_un:
+                case ILOpcode.blt_un:
+                    {
+                        var value1 = Pop();
+                        var value2 = Pop();
+
+                        CheckIsComparable(value1, value2, opcode);
+                        break;
+                    }
```

Comment:
We often add assert as default case to make sure that there is nothing failing through the cracks:

```
default:
      Debug.Assert(false, "Unexpected branch opcode");
      break;
```